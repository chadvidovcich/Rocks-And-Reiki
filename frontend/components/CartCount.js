import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';

const Dot = styled.div`
  background: var(--red);
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  margin-left: 1rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

function CartCount({ count }) {
  return (
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        classNames={count}
        className="count"
        key={count}
        timeout={{ enter: 5000, exit: 5000 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  );
}
export default CartCount;
