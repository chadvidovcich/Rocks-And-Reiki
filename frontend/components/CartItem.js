import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
    width: 10rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

export default function CartItem({ cartItem }) {
  if (!cartItem.product) return null;

  return (
    <CartItemStyles>
      <img
        src={cartItem.product.photo.image.publicUrlTransformed}
        alt={cartItem.product.name}
      />
      <div>
        <h3>{cartItem.product.name}</h3>
        <p>
          {formatMoney(cartItem.product.price * cartItem.quantity)}-
          <em>
            {cartItem.quantity} &times; {formatMoney(cartItem.product.price)}
            each
          </em>
        </p>
      </div>
    </CartItemStyles>
  );
}
