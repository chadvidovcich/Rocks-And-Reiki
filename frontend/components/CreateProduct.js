import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Router from 'next/router';
import useForm from '../lib/useForm';
import ErrorMessage from './ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';
import Form from './styles/Form';

export const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # Variables passed in with types
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      price
      description
      name
    }
  }
`;

export default function CreateProduct() {
  const { inputs, handleChange, clearForm, resetForm } = useForm({
    name: '',
    image: '',
    price: 0,
    description: '',
  });
  const [createProduct, { loading, error, data }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
      refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    }
  );
  return (
    <Form
      onSubmit={async (e) => {
        e.preventDefault();
        // submit input fields to back end
        const res = await createProduct();
        clearForm();
        // go to that product's page
        Router.push({
          pathname: `/product/${res.data.createProduct.id}`,
        });
      }}>
      <ErrorMessage error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor='image'>
          Image
          <input
            required
            type='file'
            name='image'
            id='image'
            onChange={handleChange}
          />
        </label>
        <label htmlFor='name'>
          Name
          <input
            type='text'
            name='name'
            id='name'
            placeholder='Name'
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor='price'>
          Price
          <input
            type='number'
            name='price'
            id='price'
            placeholder='Price'
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor='description'>
          Description
          <textarea
            name='description'
            id='description'
            placeholder='Description'
            value={inputs.description}
            onChange={handleChange}
          />
        </label>

        <button type='button' onClick={clearForm}>
          Clear Form
        </button>
        <button type='button' onClick={resetForm}>
          Reset Form
        </button>

        <button type='submit'>+ Add Product</button>
      </fieldset>
    </Form>
  );
}
