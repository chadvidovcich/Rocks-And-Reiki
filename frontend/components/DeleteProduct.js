import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

function update(cache, payload) {
  console.log(payload);
  console.log('running update function after delete');
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading, error }] = useMutation(
    DELETE_PRODUCT_MUTATION,
    {
      variables: {
        id,
      },
      update,
    }
  );
  return (
    <button
      type='button'
      disabled={loading}
      onClick={() => {
        // eslint-disable-next-line no-restricted-globals
        if (confirm('Are you sure you want to delete this item?')) {
          // ok to delete
          deleteProduct().catch((err) => alert(err.message));
        }
      }}>
      {children}
    </button>
  );
}
