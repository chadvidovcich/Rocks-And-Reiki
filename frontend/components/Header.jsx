// @ts-ignore
import Link from 'next/link';
import styled from 'styled-components';
import Nav from './Nav';

const Logo = styled.h1`
  font-size: 4rem;
  width: 50vw;
  margin-left: 2rem;
  position: relative;
  z-index: 2;
  /* background: var(--purple, rebeccapurple); */
  transform: skew(-12deg);
  a {
    color: var(--purple);
    text-decoration: underline;
    text-transform: uppercase;
    padding: 0.5rem 1 rem;
  }
`;

const HeaderStyles = styled.header`
  .bar {
    border-bottom: 10px solid var(--black, black);
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
  }

  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;
    border-bottom: 1px solid var(--black, black);
  }
  @media screen and (max-width: 900px) {
    .bar {
      border-bottom: 10px solid var(--black, black);
      display: grid;
      grid-template-columns: auto 1fr;
      justify-content: space-between;
      align-items: stretch;
      /* border: 10px solid green; */
    }

    .sub-bar {
      display: grid;
      grid-template-columns: 1fr auto;
      border-bottom: 1px solid var(--black, black);
    }
  }
`;

function Header() {
  return (
    <HeaderStyles>
      <div className="bar">
        <Logo>
          <Link href="/">Rocks and Reiki</Link>
        </Logo>
        <Nav />
      </div>
      <div className="sub-bar">
        <p>Search</p>
      </div>
    </HeaderStyles>
  );
}
Header.propTypes = {};
export default Header;
