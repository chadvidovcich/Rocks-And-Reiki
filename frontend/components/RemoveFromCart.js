import styled from 'styled-components';
import gql from 'graphql-tag';
// @ts-ignore
import { useMutation } from '@apollo/client';

const BigButton = styled.button`
  font-size: 4rem;
  background: none;
  border: 0;
  &:hover {
    color: var(--red);
    cursor: pointer;
  }
`;

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    deleteCartItem(id: $id) {
      id
    }
  }
`;

function RemoveFromCart({ id }) {
  const [removeFromCart, { loading }] = useMutation(REMOVE_FROM_CART_MUTATION, {
    variables: { id: id },
  });
  return (
    <BigButton
      title="Remove this Item from Cart"
      type="button"
      onClick={removeFromCart}
      disabled={loading}
    >
      &times;
    </BigButton>
  );
}
export default RemoveFromCart;
