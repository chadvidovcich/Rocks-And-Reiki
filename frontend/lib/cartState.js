import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // this is our custom provider. we will store data (state) and functionality (updates) in here. Anyone can access it via the consumer

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for accessing the cart local state
function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
