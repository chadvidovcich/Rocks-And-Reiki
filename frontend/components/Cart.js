import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import CartItem from './CartItem';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import { useUser } from './User';

export default function Cart() {
  const cartUser = useUser();
  if (!cartUser) {
    return null;
  }
  return (
    <CartStyles
      // @ts-ignore
      open
    >
      <header>
        <Supreme>{cartUser.name}'s Cart</Supreme>
      </header>
      <ul>
        {cartUser.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(cartUser.cart))}</p>
      </footer>
    </CartStyles>
  );
}
