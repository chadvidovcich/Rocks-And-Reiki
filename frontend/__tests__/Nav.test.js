// @ts-ignore
import { render, screen } from '@testing-library/react';
import { MockedProvider } from '../node_modules/@apollo/react-testing/index';
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

import { CartStateProvider } from '../lib/cartState';

// create mocks for being logged out, logged in,. and logged in with cart items.
const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: fakeUser() } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: fakeUser({
          cart: [fakeCartItem()],
        }),
      },
    },
  },
];

describe('<Nav/>', () => {
  it('renders a minimal nav when signed out', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );

    expect(container).toHaveTextContent('Sign In');
    expect(container).toMatchSnapshot();

    const link = screen.getByText('Sign In');
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/signin');

    const productsLink = screen.getByText('Products');
    expect(productsLink).toBeInTheDocument();
    expect(productsLink).toHaveAttribute('href', '/products');
  });

  it('renders a full nav when signed in', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');

    expect(container).toHaveTextContent('Sign Out');
    expect(container).toMatchSnapshot();

    const link = screen.getByText('Sign Out');
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('type', 'button');
  });

  it('renders the number of items in the cart', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocksWithCartItems}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    expect(screen.getByText('3')).toBeInTheDocument();
  });
});
