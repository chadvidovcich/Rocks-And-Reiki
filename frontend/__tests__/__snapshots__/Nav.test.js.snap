// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a full nav when signed in 1`] = `
<div>
  <ul
    class="NavStyles-sc-1w3lnmx-0 dCxQkF"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell 
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <button
      data-testid="logOutButton"
      type="button"
    >
      Sign Out
    </button>
    <button
      type="button"
    >
      My Cart
      <span
        class="CartCount__AnimationStyles-sc-1waysp3-1 kgJmyN"
      >
        <div>
          <div
            class="CartCount__Dot-sc-1waysp3-0 cXpbJa count"
          >
            0
          </div>
        </div>
      </span>
    </button>
  </ul>
</div>
`;

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<div>
  <ul
    class="NavStyles-sc-1w3lnmx-0 dCxQkF"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/signin"
    >
      Sign In
    </a>
  </ul>
</div>
`;
